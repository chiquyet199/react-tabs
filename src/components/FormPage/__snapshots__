/Users/edward/Desktop/canopy-frontend/packages/canopy-frontend-common/src/components/FormPage/__native_snapshots__