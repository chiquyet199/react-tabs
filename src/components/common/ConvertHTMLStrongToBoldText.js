import React from "react";
import {Text} from "react-native";
import styled from "styled-components/native";

export const StyledView = styled(Text)`
  display: inline;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
`;

export default ({htmlString}) => {
  const parsed = [];
  let cursor = 0;
  let searchingFor = "<strong>";
  while (cursor < htmlString.length) {
    let endPoint = htmlString.indexOf(searchingFor, cursor);
    endPoint = endPoint === -1 ? htmlString.length : endPoint;
    const token = htmlString.substring(cursor, endPoint);
    if (searchingFor === "<strong>") {
      parsed.push(<Text key={token}>{token}</Text>);
    } else {
      // Styled-components seems to ignore font-weight here
      parsed.push(
        <Text style={{fontWeight: "bold"}} key={token}>
          {token}
        </Text>,
      );
    }
    cursor = endPoint + searchingFor.length;
    searchingFor = searchingFor === "<strong>" ? "</strong>" : "<strong>";
  }

  return <StyledView>{parsed}</StyledView>;
};
