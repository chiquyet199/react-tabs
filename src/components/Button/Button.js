import styled from "styled-components/native";
import Icon from "../Icon";
import {themed} from "../../utils/common";

export const BtnText = styled.Text`
  position: relative;
  top: -1px;
  font-size: ${({type, theme: {buttons}}) => buttons[type].fontSize || "16px"};
  font-weight: ${({type, theme: {buttons}}) =>
    buttons[type].fontWeight || "normal"};
  color: ${({type, theme: {buttons}}) => buttons[type].text};
`;

export const ButtonPart = styled.View`
  justify-content: center;
  flex: 1;
`;

export const Left = styled(ButtonPart)`
  position: absolute;
  height: 46px;
  width: ${({size}) => `${size}px`};
  left: ${({shift}) => (shift ? "10px" : 0)};
  top: 0;
  bottom: 0;
`;

export const Right = styled(ButtonPart)`
  position: absolute;
  height: 46px;
  width: ${({size}) => `${size}px`};
  right: ${({shift}) => (shift ? "10px" : 0)};
  top: 0;
  bottom: 0;
`;

export const Middle = styled(ButtonPart)`
  align-items: center;
  flex: 1;
  padding: 0;
  padding-right: ${({righticon, lefticon, size}) =>
    righticon || lefticon ? `${size + 10}px` : "10px"};
  padding-left: ${({lefticon, righticon, size}) =>
    lefticon || righticon ? `${size + 10}px` : "10px"};
  min-width: 60px;
`;

const Container = styled.TouchableOpacity`
  border-width: ${props => (props.noBorder ? 0 : 1)};
  width: 100%;
  max-width: 350px;
  margin: 0;
  padding: 0;
  height: 48px;
  border-radius: 6;
  border-color: ${({type, theme: {buttons}}) => buttons[type].border};
  background-color: ${({type, theme: {buttons}}) => buttons[type].background};
  opacity: ${({disabled}) => (disabled ? 0.5 : 1)};
`;

export const ButtonContainer = styled(Container)`
  padding: 0;
  margin: 5px;
  flex: auto;
  max-height: 48px;
  width: 100%;
  align-self: center;
`;

export const StyledIcon = styled(Icon)`
  text-align: center;
`;

export const IconContainer = styled(Container)`
  width: ${({size}) => size || 48}px;
  height: ${({size}) => size || 48}px;
  margin: 0;
  padding: 0;
  align-items: center;
  justify-content: center;
`;

export const MenuButtonIconContainer = styled(IconContainer)`
  margin-top: 0;
  margin-bottom: 2px;
  background-color: transparent;
`;

export const BounceContainer = styled.View`
  position: absolute;
  width: 200px;
  margin-top: ${({theme}) => theme.padding.top}px;
  margin-right: -20px;
  right: 0;
  z-index: 999;
`;

export const ButtonIcon = themed(styled(Icon)`
  display: flex;
  font-size: ${({size}) => `${size}px` || "48px"};
  margin-bottom: ${({bias}) => bias || "0"};
  color: ${({type, theme: {buttons}}) => buttons[type].text};
`);

export const MenuButtonText = styled.Text`
  font-size: 12px;
  color: ${({type, theme: {buttons}}) => buttons[type].text};
`;

export const TabContainer = styled(ButtonContainer)`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  min-width: 142px;
  width: 100%;
  max-height: 38px;
  margin: 0;
  padding: 10px 30px;
  border-bottom: 0;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
`;
export const TabText = styled(BtnText)`
  top: 0;
  color: inherit;
`;
