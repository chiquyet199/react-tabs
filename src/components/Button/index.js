import React from "react";
import {View} from "react-native";
import {
  array,
  bool,
  func,
  object,
  oneOf,
  oneOfType,
  number,
  string,
} from "prop-types";
import {
  ButtonContainer,
  Left,
  Middle,
  Right,
  BtnText,
  StyledIcon,
  IconContainer,
  MenuButtonIconContainer,
  BounceContainer,
  MenuButtonText,
  ButtonIcon,
  TabContainer,
  TabText,
} from "./Button";
import Icon from "../Icon";
import {applyStyleObjectToStyledComponent, themed} from "../../utils/common";

export const Button = ({
  left,
  right,
  middle,
  onClick,
  type,
  noborder,
  style,
  iconsize,
  theme,
  disabled,
}) => {
  const color = theme && theme.buttons[type].text;

  return (
    <ButtonContainer
      noborder={noborder}
      type={type}
      onPress={onClick}
      activeOpacity={0.6}
      style={style}
      theme={theme}
      disabled={disabled}
    >
      {left && (
        <Left size={iconsize || 32} shift={middle}>
          <StyledIcon
            type={type}
            name={left}
            size={iconsize || 32}
            color={color}
          />
        </Left>
      )}
      {middle && (
        <Middle lefticon={left} righticon={right} size={iconsize || 32}>
          <BtnText ellipsizeMode="tail" type={type} theme={theme}>
            {middle}
          </BtnText>
        </Middle>
      )}
      {right && (
        <Right size={iconsize || 32} shift={middle}>
          <StyledIcon
            type={type}
            name={right}
            size={iconsize || 32}
            color={color}
          />
        </Right>
      )}
    </ButtonContainer>
  );
};

Button.propTypes = {
  left: oneOfType([string, bool]),
  middle: string,
  right: oneOfType([string, bool]),
  onClick: func,
  type: string,
};

Button.defaultProps = {
  left: null,
  right: null,
  type: "primary",
};

export const IconButton = ({
  onClick,
  name,
  type,
  style,
  iconsize,
  bias,
  theme,
}) => (
  <IconContainer style={style} type={type} onPress={onClick}>
    <Icon
      size={iconsize}
      type={type}
      name={name}
      bias={bias}
      color={theme && theme.buttons[type].text}
    />
  </IconContainer>
);

IconButton.propTypes = {
  name: string.isRequired,
  onClick: func,
  type: string.isRequired,
};

const ThemedIconContainer = ({
  name,
  type,
  style,
  iconsize,
  bias,
  theme: {buttons},
}) => (
  <IconContainer size={iconsize} style={style} type={type}>
    <ButtonIcon
      size={iconsize}
      type={type}
      name={name}
      bias={bias}
      color={buttons[type].text}
    />
  </IconContainer>
);

ThemedIconContainer.propTypes = {
  name: string.isRequired,
  type: string.isRequired,
  style: oneOf([array, object]),
  iconsize: number,
  bias: number,
  theme: object.isRequired,
};

export const StyledIconContainer = themed(ThemedIconContainer);

export const MenuButton = ({
  onClick,
  name,
  type,
  iconContainerStyle,
  children,
  iconsize,
  theme,
}) => {
  const StyledInnerContainer = applyStyleObjectToStyledComponent(
    iconContainerStyle,
    MenuButtonIconContainer,
  );

  return (
    <View style={{flex: 1, alignItems: "center"}}>
      <StyledInnerContainer type={type} noBorder onPress={onClick}>
        <Icon
          type={type}
          name={name}
          size={iconsize || 32}
          color={theme.buttons[type].text}
        />
      </StyledInnerContainer>
      <MenuButtonText type={type}>{children}</MenuButtonText>
    </View>
  );
};

MenuButton.propTypes = {
  onClick: func,
  name: string,
  type: string,
  iconContainerStyle: object,
  iconsize: number,
  theme: object,
};

export const FormBounceButton = ({children, onClick, globaltheme, theme}) => (
  <BounceContainer>
    <Button
      onClick={onClick}
      middle={children}
      type={theme[globaltheme].form.bounceButton}
      noBorder
    />
  </BounceContainer>
);

FormBounceButton.propTypes = {
  children: string.isRequired,
  globaltheme: oneOf(["dark", "light"]),
  onClick: func.isRequired,
};

export const TabButton = ({onClick, left, iconsize, middle, type}) => (
  <TabContainer type={type} onPress={onClick}>
    <StyledIcon type={type} name={left} size={iconsize || 20} color="inherit" />
    <TabText ellipsizeMode="tail" type={type}>
      {middle}
    </TabText>
  </TabContainer>
);

TabButton.propTypes = {
  left: string,
  type: string.isRequired,
  middle: string.isRequired,
  iconsize: number,
  onClick: func,
};

export default themed(Button);
