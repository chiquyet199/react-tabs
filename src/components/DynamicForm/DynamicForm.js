import styled from "styled-components/native/index";
import Switch from "../../components/Switch";
import {responsive, themed, isWeb} from "../../utils/common";
import DatePicker from "../DatePicker";
import {TextParagraph, PageTitle} from "../common/layout";

export const FormContainer = responsive(styled.View`
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  width: 100%;
  margin-bottom: 30px;
`)({
  S: `
  margin-bottom: 10px;
  max-width: 100%;`,
});

export const StyledInputWrapper = styled.View`
  width: 100%;
  margin-bottom: 0;
`;

export const StyledTextParagraph = styled(TextParagraph)`
  padding-left: 0;
  padding-right: 0;
`;

export const StyledPageTitle = styled(PageTitle)`
  padding-left: 0;
  padding-right: 0;
`;

export const StyledSwitch = styled(Switch)`
  margin-left: 10px;
`;

export const SwitchText = styled.Text`
  font-size: 16px;
  flex: 3;
  line-height: 24px;
  padding: 0;
  margin-right: 10px;
  color: ${({globaltheme, theme}) => theme[globaltheme].baseColor};
  text-align: ${props => (props.left ? "left" : "center")};
  align-self: ${props => (props.left ? "flex-start" : "center")};
  font-weight: ${props => (props.bold ? "bold" : "normal")};
`;

export const SwitchContainer = styled.View`
  display: flex;
  flex-direction: row;
  padding-vertical: ${({layout}) => (layout === "hq" ? "17px" : "39px")};
  align-items: center;
  border-bottom-color: ${({layout, theme}) =>
    layout === "hq" ? theme.colors.lightSteel : "rgba(255, 255, 255, 0.2)"};
  border-bottom-width: ${({layout}) => (layout === "hq" ? "1px" : "2px")};
  width: 100%;
  padding-horizontal: 0;
`;

export const SwitchError = styled(SwitchText)`
  color: ${({color, theme}) => color || theme.colors.alertRed};
`;

export const StyledInputContainer = styled.View`
  ${isWeb ? `position: static; z-index: auto;` : ""};
  display: flex;
  flex-direction: column;
  width: 100%;
`;

export const InputHeading = themed(styled.Text`
  font-size: 12px;
  font-weight: normal;
  color: ${({theme, globaltheme}) => theme[globaltheme].textColor};
  padding: 0;
`);

export const StyledHeading = themed(styled.Text`
  margin-bottom: -10px;
  padding-horizontal: 0;
  color: ${({theme, globaltheme}) => theme[globaltheme].textColor};
  font-size: 16px;
`);

export const StyledFieldWrapper = styled.View`
  ${isWeb ? `position: static; z-index: auto;` : ""};
  padding-horizontal: 5px;
  width: ${({width}) => width || 100}%;
  min-height: 92px;
`;

export const StyledDatePicker = styled(DatePicker)`
  ${isWeb ? `position: static; z-index: auto;` : ""};
  flex: 1;
  width: 100%;
`;

export const StyledPickerContainer = styled.View`
  display: flex;
  flex-direction: column;
  width: 100%;
  padding-vertical: 15px;
`;

export const ContentDivider = responsive(styled.View`
  width: ${({length}) => (length >= 0 ? `${length}px` : "100%")};
  height: 1px;
  border: 0;
  border-top-width: 1px;
  border-top-color: ${({color}) => color};
  margin: ${({margin}) => (margin >= 0 ? `${margin}px 0` : "20px 0")};
`)({
  S: `
  width: 100%;
  `,
  M: `
  width: 100%;
  `,
});

export const Tooltip = themed(
  responsive(styled.Text`
    width: ${({length}) => (length >= 0 ? `${length}px` : "160px")};
    color: ${({globaltheme, theme}) => theme[globaltheme].baseColor};
    font-size: 12px;
    position: absolute;
    top: 20px;
    left: 105%;
  `)({
    S: `
  position: relative;
  left: 0;
  width: 100%;
  `,
    M: `
  position: relative;
  left: 0;
  width: 100%;
  `,
  }),
);
