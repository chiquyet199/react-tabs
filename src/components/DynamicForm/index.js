import React from "react";
import {connect} from "react-redux";
import {TextInput} from "react-native";
import {reduxForm, Field} from "redux-form";
import {
  FormContainer,
  StyledTextParagraph,
  SwitchText,
  StyledPageTitle,
  StyledInputWrapper,
  StyledSwitch,
  SwitchContainer,
  StyledInputContainer,
  InputHeading,
  StyledHeading,
  StyledDatePicker,
  StyledFieldWrapper,
  StyledPickerContainer,
  ContentDivider,
  Tooltip,
} from "./DynamicForm";
import Picker from "../Picker";
import ImagesUpload from "../ImagesUpload";
import PostcodeLookup from "../PostcodeLookup";
import RadioButtonGroup from "../RadioButtonGroup/RadioButtonGroup";
import InputField from "../InputField/inputField";
import OptionPicker from "../OptionPicker/";
import {themed} from "../../utils/common";

// getComponent checks the schema to see if any property has a specific UI component to be rendered against it.
// For example: countryCode in the Schema has a UI component of Country-Picker, when any new schemas are added with
// new UI components to be rendered against their schema properties, a case must be added below for it's specific
// property. For functionality purposes, we've added the same component below to show the difference.
export const getComponent = ({
  initialValue,
  disabled,
  input,
  meta: {touched, error},
  schema,
  name,
  components,
  style,
  globaltheme = "light",
  theme,
  heading,
  submit,
  submitOnEnter,
  ...props
}) => {
  if (components && components[name]) {
    return components[name];
  }

  const handleKeyPress = e => {
    if (submitOnEnter && typeof submit === "function" && e.key === "Enter") {
      submit();
    }
  };

  const hasError = touched && error !== undefined;
  const {dividerColor} = theme[globaltheme];
  const {baseColor, accentColor} = theme[globaltheme].form;
  switch (schema.uiComponent) {
    case "hidden":
      return <TextInput hide="true" />;
    case "divider":
      return (
        <ContentDivider
          margin={schema.margin}
          length={schema.length}
          color={dividerColor}
        />
      );
    case "text":
      return (
        <StyledTextParagraph
          theme={theme}
          globaltheme={globaltheme}
          left={schema.align}
          bold={schema.bold}
          fontStyle={schema.fontstyle}
        >
          {schema.title}
        </StyledTextParagraph>
      );
    case "textHeader":
      return (
        <StyledPageTitle
          theme={theme}
          globaltheme={globaltheme}
          left={schema.align}
          bold={schema.bold}
          fontStyle={schema.fontstyle}
        >
          {schema.title}
        </StyledPageTitle>
      );
    case "switch":
      return (
        <SwitchContainer layout={schema.layout}>
          <SwitchText
            theme={theme}
            globaltheme={globaltheme}
            left={schema.align}
            bold={schema.bold}
            fontStyle={schema.fontstyle}
          >
            {schema.title}
          </SwitchText>
          <StyledSwitch
            disabled={disabled}
            name={input.name}
            on={!!input.value}
            onChange={input.onChange}
            onChangeCallback={schema.onChangeCallback}
          />
        </SwitchContainer>
      );
    case "password":
      return (
        <StyledInputWrapper>
          {heading && <StyledHeading>{heading}</StyledHeading>}
          <InputField
            theme={theme}
            globaltheme={globaltheme}
            tintColor={accentColor}
            textColor={baseColor}
            onChangeText={input.onChange}
            errorText={hasError && error}
            value={input.value}
            disabled={schema.disabled || false}
            autoFocus={schema.autoFocus}
            onKeyPress={handleKeyPress}
            {...props}
            password
          />
        </StyledInputWrapper>
      );
    case "imageUpload":
      return (
        <ImagesUpload
          btnText={schema.buttonText}
          btnType={schema.buttonType}
          translations={schema.translations}
        />
      );
    case "select":
      return (
        <StyledPickerContainer>
          {heading && <StyledHeading>{heading}</StyledHeading>}
          <Picker
            globaltheme={globaltheme}
            options={schema.options}
            onChange={
              schema.onChange
                ? args => {
                    schema.onChange(args);
                    input.onChange(args);
                  }
                : input.onChange
            }
            textColor={baseColor}
            borderColor={baseColor}
            currentValue={input.value}
            errorText={hasError && error}
          />
          {schema.tooltip && <Tooltip>{schema.tooltip}</Tooltip>}
        </StyledPickerContainer>
      );
    case "radio":
      return (
        <RadioButtonGroup
          layout={schema.layout}
          title={schema.title}
          options={schema.options}
          onSelect={(event, value) => input.onChange(value)}
          currentValue={input.value}
          errorText={hasError && error}
        />
      );

    case "optionPicker":
      return (
        <StyledInputContainer>
          {heading && <InputHeading>{heading}</InputHeading>}
          <OptionPicker
            globaltheme={globaltheme}
            initialValue={initialValue}
            onChange={
              schema.onChange
                ? args => {
                    schema.onChange(args);
                    input.onChange(args);
                  }
                : input.onChange
            }
            textColor={baseColor}
            borderColor={baseColor}
            type={schema.type_data}
            options={schema.options}
          />
        </StyledInputContainer>
      );

    case "datePicker":
      return (
        <StyledInputContainer>
          {heading && <InputHeading>{heading}</InputHeading>}
          <StyledDatePicker
            {...props}
            inputProps={input}
            textColor={baseColor}
            onChangeText={input.onChange}
            errorText={hasError && error}
            value={input.value}
            borderColor={baseColor}
            onChange={input.onChange}
            date={input.value}
            earliestYear={schema.earliestYear}
          />
        </StyledInputContainer>
      );
    case "postcode":
      if (!schema.onAddressSelected) {
        throw new Error("Schema must provide onAddressSelected callback");
      }
      if (!schema.onAddressNotListed) {
        throw new Error("Schema must provide onAddressNotListed callback");
      }

      return (
        <PostcodeLookup
          onAddressSelected={schema.onAddressSelected}
          onAddressNotListed={schema.onAddressNotListed}
          onChange={input.onChange}
        />
      );
    default:
      return (
        <StyledInputWrapper>
          {heading && <StyledHeading>{heading}</StyledHeading>}
          <InputField
            {...props}
            tintColor={accentColor}
            textColor={baseColor}
            onChangeText={input.onChange}
            errorText={hasError && error}
            defaultValue={input.value}
            keyboardType={schema.keyboardType}
            multiline={schema.multiline}
            heading={schema.heading}
            prefix={schema.prefix}
            disabled={schema.disabled || false}
            theme={theme}
            autoFocus={schema.autoFocus}
            onKeyPress={handleKeyPress}
            autoCapitalize={
              schema.keyboardType === "email-address" ? "none" : "sentences"
            }
          />
        </StyledInputWrapper>
      );
  }
};

export const GenerateFields = formProps => {
  const {
    initialValues = {},
    schema,
    globaltheme,
    theme,
    components,
    submit,
    submitOnEnter,
  } = formProps;

  return Object.entries(schema.properties).map(([key, object]) => {
    const props = {
      name: key,
      globaltheme,
      theme,
      heading: object.heading,
      component: getComponent,
      initialValue: initialValues[key],
      disabled: object.disabled,
      props: {
        schema: object,
        name: key,
        components,
        submit,
        submitOnEnter,
      },
      label: object.title,
    };

    return (
      <StyledFieldWrapper width={object.width} key={key}>
        <Field {...props} />
      </StyledFieldWrapper>
    );
  });
};

export const DynamicForm = props => (
  <FormContainer>
    <GenerateFields {...props} />
  </FormContainer>
);

const connectedForm = reduxForm({
  form: "DynamicForm",
  enableReinitialize: true,
  keepDirtyOnReinitialize: true,
  destroyOnUnmount: false,
})(DynamicForm);

const mapStateToProps = (state, props) => ({
  initialValues: props.initialValues,
});

const StateForm = connect(
  mapStateToProps,
  {},
)(themed(connectedForm));

export default StateForm;
