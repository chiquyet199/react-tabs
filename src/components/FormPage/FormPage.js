import React, {Component, Fragment} from "react";
import {connect} from "react-redux";
import DynamicForm from "../DynamicForm";
import Button from "../Button";
import {
  PageTitle,
  PageContainer,
  TextParagraph,
  ErrorMessage,
  ScrollContentContainer,
  ScrollButtonBar,
  ScrollScrollView,
} from "../common/layout";
import {SubmissionError} from "redux-form";
import {isEmpty} from "underscore";
import {validate} from "../../utils/validators/common";
import {isWeb} from "../../utils/common";
import {object, bool, string, shape} from "prop-types";

export class BaseFormPage extends Component {
  state = {
    displayMultiSwitchError: false,
  };

  withValidation = parentSubmit => async values => {
    const {schema, userType = "admin", inviteUser, translations} = this.props;
    const errors = await validate(schema.properties, values);
    if (!isEmpty(errors)) {
      /*
        `branch_` indicates multiple switches
        check if user selected at least one of them
      */
      if (Object.keys(errors).some(item => item.includes("branch_"))) {
        await this.setState({
          displayMultiSwitchError: true,
        });
      }

      Object.keys(errors).forEach(error => {
        errors[error] = translations.validationErrors[errors[error]];
      });

      throw new SubmissionError(errors);
    }

    if (schema.customInviteUserValidation && userType !== "user") {
      await inviteUser(values);
    }

    return parentSubmit(values);
  };

  renderHeaderAndInfo(schema) {
    const headerAndInfo = [];

    if (schema.header) {
      headerAndInfo.push(
        <PageTitle
          left="true"
          key={`page_header_${schema.header.toLowerCase().replace(" ", "_")}`}
        >
          {schema.header}
        </PageTitle>,
      );
    }
    if (schema.info) {
      headerAndInfo.push(
        <TextParagraph
          left="true"
          key={`page_info_${schema.info.toLowerCase().replace(" ", "_")}`}
        >
          {schema.info}
        </TextParagraph>,
      );
    }

    return headerAndInfo;
  }

  componentDidUpdate(prevProps) {
    if (prevProps.hasErrors !== this.props.hasErrors && this.props.hasErrors) {
      this.props.submit();
    }
  }

  render() {
    const {displayMultiSwitchError} = this.state;
    const {
      schema,
      components,
      omitSubmit,
      submit,
      onSubmit,
      submitOnEnter,
      screenSize,
      multiSwitchErrorText,
      extraButtons,
      extraButtonBarButtons,
    } = this.props;

    if (schema) {
      const initialValues = this.props.initialValues || schema.initialValues;
      const {
        buttonStyles,
        buttonIcon = "arrow-right",
        buttonText = "Next",
        buttonType = "primary",
        layout,
        marginTop,
      } = schema;

      let multiSwitchError = null;

      if (displayMultiSwitchError) {
        multiSwitchError = <ErrorMessage>{multiSwitchErrorText}</ErrorMessage>;
      }

      return (
        <ScrollContentContainer size={screenSize}>
          <ScrollScrollView>
            <PageContainer layout={layout} margintop={marginTop}>
              {this.renderHeaderAndInfo(schema)}
            </PageContainer>
            <PageContainer id="dynamicFormWrapper">
              <DynamicForm
                initialValues={initialValues}
                schema={schema}
                components={components}
                onSubmit={this.withValidation(onSubmit)}
                submit={submit}
                submitOnEnter={submitOnEnter}
              />
            </PageContainer>
          </ScrollScrollView>
          {multiSwitchError}
          {!omitSubmit && (
            <Fragment>
              <ScrollButtonBar style={buttonStyles}>
                <Button
                  type={buttonType}
                  right={buttonIcon}
                  middle={buttonText}
                  onClick={submit}
                />
                {extraButtonBarButtons && extraButtonBarButtons}
              </ScrollButtonBar>
              <ScrollButtonBar style={buttonStyles}>
                {extraButtons && extraButtons}
              </ScrollButtonBar>
            </Fragment>
          )}
        </ScrollContentContainer>
      );
    }

    return null;
  }
}

export const FormPage = props => (
  <BaseFormPage {...props}>
    {isWeb ? (
      <form>{props.children}</form>
    ) : (
      <Fragment>{props.children}</Fragment>
    )}
  </BaseFormPage>
);

FormPage.defaultProps = {
  omitSubmit: false,
};

FormPage.propTypes = {
  hasErrors: bool,
  schema: object.isRequired,
  omitSubmit: bool,
  userType: string,
  multiSwitchErrorText: string,
  extraButtonBarButtons: object,
  extraButtons: object,
  submitOnEnter: bool,
  translations: shape({
    validationErrors: shape({
      onlyLetters: string.isRequired,
      required: string.isRequired,
      selectAtLeastOne: string.isRequired,
      invalidEmail: string.isRequired,
      emailTaken: string.isRequired,
      onlyNumbers: string.isRequired,
      invalidNumber: string.isRequired,
      invalidPassword: string.isRequired,
      passwordsDontMatch: string.isRequired,
      invalidPhoneNumber: string.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = state => ({
  multiSwitchErrorText: state.locale.translations.common.multiSwitchError,
  translations: {
    validationErrors: state.locale.translations.common.validationErrors,
  },
});

export default connect(
  mapStateToProps,
  null,
)(FormPage);
