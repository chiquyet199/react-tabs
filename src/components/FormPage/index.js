import {connect} from "react-redux";
import {submit} from "redux-form";
import FormPage from "./FormPage";
import {go} from "../../actions/navigation";

export const mapStateToProps = state => ({
  screenSize: state.dimensions.size,
});

const mapDispatchToProps = {
  submit: () => dispatch => dispatch(submit("DynamicForm")),
  go,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FormPage);
