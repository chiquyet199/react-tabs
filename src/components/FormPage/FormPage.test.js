import React from "react";
import {BaseFormPage} from "./FormPage";
import {shallow} from "enzyme";
import {
  ScrollButtonBar,
  PageContainer,
  ErrorMessage,
  ScrollContentContainer,
} from "../common/layout";
import themes from "../../constants/theme";
import {extend} from "underscore";
import {validate} from "../../utils/validators/common";
import DynamicForm from "../DynamicForm";
const layout = "hq";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import {initialState} from "../../__mocks__/initialState";
import Button from "../Button";
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const theme = themes.selva;
jest.mock("../../utils/validators/common");

describe("FormPage tests", () => {
  it("renders to correct snapshot", () => {
    expect(getComponent()).toMatchSnapshot();
  });

  it("Pass initial values to Dynamic Form", () => {
    const initialValues = {info: "test"};
    const component = getComponent({initialValues});
    expect(component.find(DynamicForm).prop("initialValues")).toEqual(
      initialValues,
    );
  });

  it("Shows button bar if omitSubmit property set to false", () => {
    const component = getComponent({omitSubmit: false});
    expect(component.find(ScrollButtonBar)).toExist();
  });

  it("Hides button bar if omitSubmit property set to true", () => {
    const component = getComponent({omitSubmit: true});
    expect(component.find(ScrollButtonBar)).not.toExist();
  });

  it("Sets button type when provided in schema", () => {
    const schema = mockSchema();
    schema.buttonType = "big, red button";
    const component = getComponent({schema});
    expect(component.find(Button)).toHaveProp("type", schema.buttonType);
  });

  describe("when `layout` prop is passed", () => {
    it("applies appropriate prop to the `PageContainer`", () => {
      const component = getComponent();
      const container = component
        .find(PageContainer)
        .findWhere(node => node.prop("layout") === "hq");
      expect(container).toHaveLength(1);
    });
  });

  describe("when `size` prop is passed", () => {
    it("applies appropriate prop to the `ScrollContentContainer`", () => {
      const screenSize = "S";
      const component = getComponent({screenSize});
      const sizeProp = component.find(ScrollContentContainer).prop("size");
      expect(sizeProp).toBe(screenSize);
    });
  });

  describe("displayMultiSwitchError", () => {
    it("displays an error when `DynamicForm` encountered `branch_` error", async () => {
      validate.mockImplementation(() => {
        const branchesErrors = {
          branch_Islington_permission: "Please select at least 1",
        };

        return Promise.resolve(branchesErrors);
      });
      const component = getComponent();
      const onSubmit = component.find(DynamicForm).prop("onSubmit");
      try {
        await onSubmit(() => {}, {branch_Islington_permission: false});
      } catch (e) {
        // we're expecting a SubmissionError, we need to swallow it so the test can continue...
      } finally {
        component.update();
        expect(component.find(ErrorMessage)).toExist();
      }
    });
  });

  const getComponent = props => {
    const parsedProps = extend(
      {
        schema: mockSchema(),
        omitSubmit: false,
        globaltheme: "light",
        theme,
        translations: {
          validationErrors: {
            onlyLetters: "onlyLetters",
            required: "required",
            selectAtLeastOne: "selectAtLeastOne",
            invalidEmail: "invalidEmail",
            emailTaken: "emailTaken",
            onlyNumbers: "onlyNumbers",
            invalidNumber: "invalidNumber",
            invalidPassword: "invalidPassword",
            passwordsDontMatch: "passwordsDontMatch",
            invalidPhoneNumber: "invalidPhoneNumber",
          },
        },
      },
      props,
    );

    return shallow(<BaseFormPage {...parsedProps} />, {
      context: {store: mockStore(initialState)},
    });
  };
});

const mockSchema = () => ({
  layout,
  progress: 33,
  next: "password",
  type: "object",
  title: "about",
  header: "About you",
  info: "Test info",
  initialValues: {
    firstName: "Benedict",
    lastName: "Cumberland",
  },
  properties: {
    info: {
      type: "string",
      title: "A title",
      uiComponent: "text",
      align: "left",
    },
    firstName: {
      type: "string",
      title: "First name",
    },
    lastName: {
      type: "string",
      title: "Last name",
    },
    email: {
      type: "string",
      title: "address",
    },
    receive_emails: {
      type: "boolean",
      title: "emails",
      uiComponent: "switch",
    },
  },
  required: ["firstName", "lastName", "email"],
});
