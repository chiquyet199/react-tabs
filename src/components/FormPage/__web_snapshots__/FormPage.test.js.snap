// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormPage tests renders to correct snapshot 1`] = `
<Styled(View)>
  <Styled(Component)>
    <Connect(ResponsiveWrapper)
      layout="hq"
    >
      <Connect(WithTheme(Styled(Text)))
        key="page_header_about_you"
        left="true"
      >
        About you
      </Connect(WithTheme(Styled(Text)))>
      <Connect(WithTheme(Styled(Text)))
        key="page_info_test_info"
        left="true"
      >
        Test info
      </Connect(WithTheme(Styled(Text)))>
    </Connect(ResponsiveWrapper)>
    <Connect(ResponsiveWrapper)
      id="dynamicFormWrapper"
    >
      <Connect(Connect(WithTheme(ReduxForm)))
        initialValues={
          Object {
            "firstName": "Benedict",
            "lastName": "Cumberland",
          }
        }
        onSubmit={[Function]}
        schema={
          Object {
            "header": "About you",
            "info": "Test info",
            "initialValues": Object {
              "firstName": "Benedict",
              "lastName": "Cumberland",
            },
            "layout": "hq",
            "next": "password",
            "progress": 33,
            "properties": Object {
              "email": Object {
                "title": "address",
                "type": "string",
              },
              "firstName": Object {
                "title": "First name",
                "type": "string",
              },
              "info": Object {
                "align": "left",
                "title": "A title",
                "type": "string",
                "uiComponent": "text",
              },
              "lastName": Object {
                "title": "Last name",
                "type": "string",
              },
              "receive_emails": Object {
                "title": "emails",
                "type": "boolean",
                "uiComponent": "switch",
              },
            },
            "required": Array [
              "firstName",
              "lastName",
              "email",
            ],
            "title": "about",
            "type": "object",
          }
        }
      />
    </Connect(ResponsiveWrapper)>
  </Styled(Component)>
  <Styled(Styled(View))>
    <Connect(WithTheme(Button))
      middle="Next"
      right="arrow-right"
      type="primary"
    />
  </Styled(Styled(View))>
  <Styled(Styled(View)) />
</Styled(View)>
`;
